### ISSUE TITLE
Missing support for many trivial builtin functions in operator

### ISSUE THREAD

**Issue by Skylion007** [pytorch/pytorch#116396]
### 🐛 Describe the bug

We are missing support for a few operators, operators.iconcat and operator.concat 
```python
import functools
import operator
import torch

def fn_works() -> None:
    a = list(range(1, 3))
    b = list(range(3, 6))
    c = [a, b]
    d = functools.reduce(operator.add, c, [])
    e = functools.reduce(operator.iadd, c, [])

def fn() -> None:
    a = list(range(1, 3))
    b = list(range(3, 6))
    c = [a, b]
    d = functools.reduce(operator.concat, c, [])
    e = functools.reduce(operator.iconcat, c, [])

comp_out_works = torch._dynamo.optimize(nopython=True)(fn_works)()
comp_out = torch._dynamo.optimize(nopython=True)(fn)()
```

Just glancing at operators documentation and the current builtin dynamo code: https://github.com/pytorch/pytorch/blob/c7e9c1510274184b41e408e6409f252bb1717085/torch/_dynamo/variables/builtin.py#L92

Things that work that I did not expect:
* operator.is
* operator.is_not
* operator.__add__ # hurray for aliasing

I think we are also missing support for:
* operator.concat
* operator.iconcat
* operator.contains # almost certainly broken. Be careful this when has reversed operands
* operator.countOf
* operator.delitem # might make sense not to fold?
* operator.getitem
* operator.setitem # might make sense not to fold?
* operator.abs # surprisingly broken given builtins.abs works
* operator.length_hint # confirmed broken 
* operator.call # python 3.11+ only
* operator.attrgetter
* operator.itemgetter
* operator.methodcaller 

```python
def fn_abs() -> None:
    a = range(-10, 10)
    b = map(operator.abs, a)

comp_out = torch._dynamo.optimize(nopython=True)(fn_abs)()
```

@XuehaiPan want to take a look at this? You did a great job fixing some of the already broken ones.

### Versions

Tried on latest stable, but I am sure it's on master too from looking at the source.

**Comment by XuehaiPan** [comment]
We can add `operator.abs` to:

https://github.com/pytorch/pytorch/blob/c7e9c1510274184b41e408e6409f252bb1717085/torch/_dynamo/variables/builtin.py#L89-L147

We can ignore the dunder version, such as `operator.__pos__`, because they are function aliases. These unary functions can calculate the results if the input is a constant Python variable.

For `operator.cancat`, it only supports sequences:

```python
def concat(a: Sequence[T], b: Sequence[T]) -> Sequence[T]:
    "Same as a + b, for a and b sequences."
    if not hasattr(a, '__getitem__'):
        msg = "'%s' object can't be concatenated" % type(a).__name__
        raise TypeError(msg)
    return a + b
```

We should handle `ConstVariable(str)` and sequence-like variable containers (e.g., `ListVariable`, `TupleVariable`, `DequeVariable`) differently.

**Comment by Skylion007** [comment]
True, although raising that TypeError shouldn't be a huge blocker. We just need to add it to the shim to this dictionary, right? https://github.com/pytorch/pytorch/blob/c7e9c1510274184b41e408e6409f252bb1717085/torch/_dynamo/variables/builtin.py#L362

**Comment by Skylion007** [comment]
So `operator.is` and `operator.is_not` seems to have a fun bug I found by trying to enable more of the test suite... Hurray... https://github.com/pytorch/pytorch/pull/116397 not even sure where to start looking to track this down.

**Comment by exitflynn** [comment]
Hello, there! i've been wanting to find a way to start contributing for a while now and I'd love to try and make progress on this issue!
I've got some general understanding of `dynamo` and shim. Could I get some pointers? Or maybe be linked to similar PR's related to this one before which I can look at for reference, to better understand the problem and the fixes?

**Comment by VimalN2005** [comment]
Hi maintainers,

I investigated this issue against the latest `main` branch and wanted to share the current status:

1. **`operator.concat` and `operator.iconcat`:**
   Both are already implemented in TorchDynamo via polyfills:
   - `operator.concat` is defined in [`torch/_dynamo/polyfills/operator.py`](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/polyfills/operator.py#L111-L115) with `@substitute_in_graph(operator.concat, can_constant_fold_through=True)`.
   - `operator.iconcat` is defined in [`torch/_dynamo/polyfills/operator.py`](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/polyfills/operator.py#L123-L127) with `@substitute_in_graph(operator.iconcat)`.
   - Corresponding unit tests are in [`test/dynamo/test_functions.py`](https://github.com/pytorch/pytorch/blob/main/test/dynamo/test_functions.py#L4350-L4368) (`test_operator_concat` and `test_operator_iconcat`).

2. **Remaining `operator` functions:**
   While `concat`/`iconcat` are covered, a few other operators (such as `operator.matmul` and `operator.imatmul`) are not currently polyfilled in `torch/_dynamo/polyfills/operator.py`.

If `operator.concat` / `operator.iconcat` was the primary focus here, this issue appears resolved and can be closed. If you'd like the remaining operators (like `operator.matmul`) polyfilled as well, I'd be happy to submit a quick PR for those!
